### DISCUSSION TITLE
How to speed up opening drawer host in Material Design in xaml toolkit?

### DISCUSSION THREAD

**Discussion by VahidEra** [MaterialDesignInXAML/MaterialDesignInXamlToolkit#3037] [Q&A]
Hi guys,
I have a drawer host in my WPF project and I want to increase its opening speed because by default its opening speed is slow. Is there any chance to increase its opening speed?

***This is my cod:***
```
      <materialDesign:DrawerHost 
            OpenMode="Modal"
            x:Name                    ="draPlaylist" 
            IsRightDrawerOpen          ="{Binding DrawerOpen}"   
            RightDrawerBackground       ="White"         
            RightDrawerCloseOnClickAway ="True" 
            Foreground                ="Black" 
            RenderTransformOrigin     ="0.5,0.5" Margin="0,0,-1,-1">
            <materialDesign:DrawerHost.RightDrawerContent>
                <Grid>
                    <Grid.RowDefinitions>
                        <RowDefinition Height="Auto"/>
                        <RowDefinition Height="Auto"/>
                    </Grid.RowDefinitions>
                <StackPanel
                    Grid.Row="0"
                    Width="200"                   
                    x:Name="stPlaylist">
                </StackPanel>
                </Grid>
            </materialDesign:DrawerHost.RightDrawerContent>
        </materialDesign:DrawerHost>
```

> **I have no idea how to do this.**

**Comment by nicolaihenriksen** [comment]
@VahidEra I think you will have a hard time increasing the speed of the animation without having to write a lot of stuff on your own. The animations are defined in [Generic.xaml](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/3103d0775618db1f19c31c2656a0e20bc9a96a31/MaterialDesignThemes.Wpf/Themes/Generic.xaml#L673-L689) (the linked example shows the left drawer).

You do have the option to disable the animation though. You can do that by setting `TransitionAssist.DisableTransitions="True"` on your `DrawerHost`. This is an inherited attached property, so you may want to enable it again on a child element if you want transitions inside of your drawer.

**Reply by VahidEra** [reply]
@nicolaihenriksen Dear friend,
You are absolutely right and as always you answered my question completely and flawlessly. I wish Material Design would provide more options to the programmer so that he could meet his needs without manipulating the original code. :-)

**Reply by nicolaihenriksen** [reply]
> @nicolaihenriksen Dear friend, You are absolutely right and as always you answered my question completely and flawlessly. I wish Material Design would provide more options to the programmer so that he could meet his needs without manipulating the original code. :-)

I believe the intention to expose properties to give the user a lot of control is definitely there. And creating a discussion topic or an issue in the repository is definitely a good way to indicate the desire for something which might be missing.

Having an option to control the duration of this animation could be a relevant addition. However, if I remember correctly, I believe binding to an attached property (which is how we typically expose hooks) from within a storyboard transition is problematic; typically it throws a binding error like _"Cannot find governing FrameworkElement or FrameworkContentElement for target element"_. @Keboo do you have any great ideas on how to expose such a duration for the user to manipulate it? I guess it should also be considered whether it is a "global transition duration" similar to `TransitionAssist.DisableTransitions`, or there should be hooks for all the different animations in the library (whoa, that could be quite a big task)...

**Reply by Keboo** [reply]
Yea the storyboards and bindings don't really work.
From the [docs](https://learn.microsoft.com/dotnet/desktop/wpf/graphics-multimedia/storyboards-overview?redirectedfrom=MSDN&view=netframeworkdesktop-4.8#animate-in-a-style)

> You can't use dynamic resource references or data binding expressions to set [Storyboard](https://learn.microsoft.com/en-us/dotnet/api/system.windows.media.animation.storyboard) or animation property values. That's because everything inside a [Style](https://learn.microsoft.com/en-us/dotnet/api/system.windows.style) must be thread-safe, and the timing system must [Freeze](https://learn.microsoft.com/en-us/dotnet/api/system.windows.freezable.freeze)[Storyboard](https://learn.microsoft.com/en-us/dotnet/api/system.windows.media.animation.storyboard) objects to make them thread-safe. A [Storyboard](https://learn.microsoft.com/en-us/dotnet/api/system.windows.media.animation.storyboard) cannot be frozen if it or its child timelines contain dynamic resource references or data binding expressions.
 
It also gets complicated since we probably want to expose a simple API like `TimeSpan` but many of the target properties will need to be converted to various other target types. 

Though in this case I think we have some options.

When doing the bindings 

Another thought I have would be to make static properties that could be set in code (likely need a new class to hold them). This would allow a measure of configuration, but it would, at best, be a per-style level configuration (you would not be able to change it for a single instance of a control). This might be good enough for most people, as I suspect if you are modifying the durations, you likely don't need (or want) different durations per instance.
It would end up looking something like this (this is just a proof of concept, I am not sure I like the API).
![image](https://user-images.githubusercontent.com/952248/211489903-c3248f8b-30c6-464d-9da0-958b569d4c04.png)
![image](https://user-images.githubusercontent.com/952248/211489993-b718f1bd-2575-4dee-b1c0-afad89928ec3.png)
Then to change the animation speed, consumers could do something like this in their own app.
![image](https://user-images.githubusercontent.com/952248/211490078-dd3cc3ae-98bb-4cca-8ecf-a03874daf523.png)

The big drawbacks I see with this approach are:
1. Set once per-style, can't change per-instance
2. Nested styles may be problematic and cause un-intended consequences. Imaging the ripple animation, and wanting different speeds for the one in the button vs a checkbox.

Another option we could consider is exposing extension methods on the control types that really just handle the lookup that a caller _could_ do today. Something like this:
![image](https://user-images.githubusercontent.com/952248/211496586-ff651260-5252-44fd-a3b4-9214d7956ff1.png)

This approach would give per-instance level control, but it comes with a bunch of negatives:
1. Pain to invoke correctly. Because it relies on looking up the visual states it needs to be called in a Loaded event handler or similar timing. We might be able to smooth over some of this is a behavior or similar.
2. Must be called form code. Again, we might be able to do something fancy with a behavior (maybe put some logic in the on-changed of an attached property or similar) to try and provide a nicer interface, but still not great.
3. Bit of a pain to maintain, though with some of the source generation work I want to do for 5.x perhaps we could generate it?

Other thoughts people have?

**Reply by nicolaihenriksen** [reply]
@Keboo Thanks for the input!

The approach with the static properties, to me, seems like the better of the two options. I feel the second option creates an even tighter coupling between the assist-class and the XAML than we had before; The assist-class now needs to have implementation knowledge about the storyboards, transitions, etc. from the corresponding XAML.

For your second bullet in the static properties approach: Perhaps this could be partly mitigated by adding the `RippleAnimationTime` on the context-specific assist types rather than `RippleAssist`. I know this is not the way we "normally do it today", but having a `ButtonAssist.RippleAnimationTime`, `CheckBoxAssist.RippleAnimationTime`, etc. would be one way to get around this?! It would still be a once-per style, but you would have a little more flexibility
